pkg/submitter/submitter.go

```diff
@@ -0,0 +1,510 @@
+		select {
+		case <-ctx.Done(): // The context was canceled from the Start loop because new work arrived.
+			return
+		case <-timeToSubmit.Done(): // It is time to submit so start the submit loop
+			for {
+				// There is no new challenge so resend any pending solution.
+				if result == nil {
+					if s.solutionPending == nil {
+						return
+					}
+					result = s.solutionPending
+					var ids []int64
+					for _, id := range s.solutionPending.Work.Challenge.RequestIDs {
+						ids = append(ids, id.Int64())
+					}
+					level.Debug(s.logger).Log("msg", "re-submitting a pending solution", "reqIDs", fmt.Sprintf("%+v", ids))
+				}
+				// Set this solution as pending so that if
+				// any of the checks below fail and will be retried
+				// when there is no new challenge.
+				s.solutionPending = result
+
+				profitPercent, err := s.profit() // Call it regardless of whether we use so that is sets the exposed metrics.
+				if s.cfg.Mine.ProfitThreshold > 0 {
+					if err != nil {
+						level.Error(s.logger).Log("msg", "submit solution profit check", "err", err)
+					} else if profitPercent < int64(s.cfg.Mine.ProfitThreshold) {
+						level.Debug(s.logger).Log("msg", "transaction not profitable, so will wait for the next cycle")
+					} else { // transaction is profitable.
+						tx, err := s.Submit(s.ctx, result)
+						if err != nil {
+							level.Error(s.logger).Log("msg", "submiting a solution", "err", err, "account", s.account.Address.String())
+							s.submitFailCount.Inc()
+							return
+						}
+						level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String(), "account", s.account.Address.String())
+						s.saveGasUsed(ctx, tx)
+						s.submitCount.Inc()
+						// A solution has been submitted so the
+						// pending solution doesn't matter here any more so reset it.
+						s.solutionPending = nil
+						return
+					}
+				}
+
+				// At this point it means that profit was not enough so need to retry in some time.
+				level.Debug(s.logger).Log("msg", "profit was not enough, retry in 60 seconds", "account", s.account.Address.String())
+				retryTicker := time.NewTicker(60 * time.Second)
+				select {
+				case <-ctx.Done(): // for any blocking operation always check the main context for an early exit.
+				case <-retryTicker.C:
+					continue // Retry send.
+				}
+			}
+		}
+	}(ctx, result)
+}
+
+func (s *Submitter) Submit(ctx context.Context, result *mining.Result) (*types.Transaction, error) {
+	challenge := result.Work.Challenge
+	nonce := result.Nonce
+	s.currentChallenge = challenge
+	s.currentNonce = nonce
+
+	// The submit contains values for 5 data IDs so add them here.
+	for i := 0; i < 5; i++ {
+		valKey := fmt.Sprintf("%s%d", db.QueriedValuePrefix, challenge.RequestIDs[i].Uint64())
+		m, err := s.proxy.BatchGet([]string{valKey})
+		if err != nil {
+			return nil, errors.Wrapf(err, "retrieve pricing data for current request id")
+		}
+		val := m[valKey]
+		var value *big.Int
```

Comment:
If we are getting the values form the db here then the tasker doesn't need to do the same.

At the moment the tasker also does the same.
```suggestion
		valKey := fmt.Sprintf("%s%d", db.QueriedValuePrefix, challenge.RequestIDs[i].Uint64())
		m, err := s.proxy.BatchGet([]string{valKey})
		if err != nil {
			return nil, errors.Wrapf(err, "retrieve pricing data for current request id")
		}
		val := m[valKey]
		var value *big.Int
```

Reply:
I think that we check values in the tasker, and then again we set those values in the submitter. am I wrong?

Reply:
Why do we need to check them in the tasker?

Reply:
Maybe to check to see if the pricing data is available for the request and return nil if it isn't.

Reply:
Hm, I think this should only happen in the submitter.
If pricing data is not available, it is set to pending and can be retried in 15sec.

Reply:
Thanks, I will do that.